druid/src/widget/rotated.rs

```diff
@@ -0,0 +1,111 @@
+use crate::event::Event::{MouseDown, MouseMove, MouseUp, Wheel};
+use crate::{
+    Affine, BoxConstraints, Env, Event, EventCtx, LayoutCtx, LifeCycle, LifeCycleCtx, MouseEvent,
+    PaintCtx, RenderContext, Size, UpdateCtx, Vec2, Widget,
+};
+use std::f64::consts::PI;
+
+pub struct Rotated<W> {
+    child: W,
+    quarter_turns: u8,
+    transforms: Option<(Affine, Affine)>,
+}
+
+impl<W> Rotated<W> {
+    pub fn new(child: W, quarter_turns: u8) -> Self {
+        Rotated {
+            child,
+            quarter_turns,
+            transforms: None,
+        }
+    }
+}
+
+impl<W> Rotated<W> {
+    fn flip(&self, size: Size) -> Size {
+        if self.quarter_turns % 2 == 0 {
+            size
+        } else {
+            Size::new(size.height, size.width)
+        }
+    }
+
+    fn affine(&self, child_size: Size, my_size: Size) -> Affine {
+        let a = ((self.quarter_turns % 4) as f64) * PI / 2.0;
+
+        Affine::translate(Vec2::new(my_size.width / 2., my_size.height / 2.))
+            * Affine::rotate(a)
+            * Affine::translate(Vec2::new(-child_size.width / 2., -child_size.height / 2.))
+    }
+
+    fn translate_mouse_event(&self, inverse: Affine, me: &MouseEvent) -> MouseEvent {
+        let mut me = me.clone();
+        me.pos = inverse * me.pos;
+        me
+    }
+}
+
+impl<T, W: Widget<T>> Widget<T> for Rotated<W> {
+    fn event(&mut self, ctx: &mut EventCtx, event: &Event, data: &mut T, env: &Env) {
+        if let Some((transform, inverse)) = self.transforms {
+            ctx.widget_state.invalid.transform_by(inverse);
```

Comment:
Every context with a `request_paint` method can modify the invalid region, so this also needs to be in `update` and `lifecycle`. But I think it would actually be nicer (and maybe less fragile) to handle this in `WidgetState`. That is, replace `WidgetState::viewport_offset` by, say, `WidgetState::viewport_transform`,(which would be an `Affine`), and then handle the transformation of invalid regions during `WidgetState::merge_up`.

Reply:
Yes - as I mentioned in the description,  I didn't want to do too much with this region stuff before the general idea was thought to make sense.